### ISSUE TITLE
How to quantize weights in forward pass during training?

### ISSUE THREAD

**Issue by mattroos** [fchollet/keras#4070]
I'd like to train a network with binary weights in the manner of Coubariaux, et al. (https://arxiv.org/abs/1602.02830), but I can't figure out where the quantization (binarization) should occur within the code.

A core aspect of the Courbariaux training method is this:
- At the beginning of each batch during training, the stored real (e.g., int32 or float32) weights are converted to binary values*\* (either by rounding or in a stochastic/probabilistic manner) and stored separately from the real-valued weights.
-  Binary-valued weights are used in the forward pass, to compute activations.
-  Real-valued weights are used in the backward pass, to compute the gradients.
-  Weight updates are applied to the real-valued weights, not the binary ones.
-  Binary-valued weights are not altered until the next batch, at which time they are recomputed by binarizing the recently updated real-valued weights.

If I do the binarization in Layer.call(), I believe it will occur for every forward pass (for every sample), but it should only occur once per batch.

If I do the binarization in a Callback.on_batch_begin(), I don't think I can specify the use of binary weights for the forward pass, and real-valued weights for gradient computation.

Any suggestions? Unfortunately my Python knowledge is not deep, so it's been somewhat challenging for me to understand the flow of the code.

**Note that when I talk about binary values, I do not mean 1-bit (e.g., 8 binary values placed into an int8). The binary-valued weights can still be represented by int32, float32, whatever.

**Comment by VanushVaswani** [comment]
There needs to be backend support. Check https://github.com/fchollet/keras/issues/2019

**Comment by mattroos** [comment]
Thanks, Vanush, but I don't believe that to be true in the context of the specific goal I described.  I'm not looking to use the binary values to save memory or accelerate processing.  This is for research purposes so keeping the binary values stored as int32 or float32 is fine (for now).

I'm using the Theano backend.  Courbariaux et al., implemented [training of binary weights in lasagne](https://github.com/MatthieuCourbariaux/BinaryNet) in the manner I described.  True, they also built a GPU binary kernel that can be used for forward passes, but they don't use it for training the network weights.

I can believe changes or additions will be needed in theano_backend.py, but no changes to Theano should be needed.

**Comment by yliapis** [comment]
I am actually trying to do the same thing right now. I want to be able to have different behavior between the forward and backward pass during training. Is there a clean way to do this?

**Comment by mattroos** [comment]
I never found a solution in Keras.  I've mostly moved on to Pytorch for research-grade NNs.

**Comment by lgeiger** [comment]
@mattroos @yliapis you can checkout http://larq.dev and https://github.com/larq/larq with `tf.keras` based implementations for binary and extremely quantized models.
